Fix 10517 - Prevent tokens without addresses from being added to token list

darkwing opened the pull request:
Fixes: #10517

Explanation:  

ToDo:
* [x] Add migration
* [x] Throw error in `addToken` to prevent this from happening again
* [x] Explore `ui/app/ducks/swaps/swaps.js` to see if this is truly the place it's happening, and if `selectedFromToken` and `selectedToToken` are the true culprits.

darkwing commented on ui/app/ducks/swaps/swaps.js:
```diff
@@ -429,7 +429,11 @@ export const fetchQuotesAndSetQuoteState = (
     let destinationTokenAddedForSwap = false;
-    if (toTokenSymbol !== 'ETH' && !contractExchangeRates[toTokenAddress]) {
+    if (
+      toTokenAddress &&
```
This is necessary because the next two lines would allow `const selectedToToken =  {}` to add a token, which we definitely don't want.

danjm replied:
This looks like a fine change to protect against errors. However, we should never be in a situation where there is no `toTokenAddress`. If so, requesting quotes should fail.

In the issue description, it was described how editing the code so that `toTokenAddress` is undefined results in the described bug, which occurs after confirming the swap. So it sounds like requesting quotes succeeded (otherwise there wouldn't be a quote to submit on the view quote screen).

So that leads me to believe that there is some other value that is being used for the token address in the request for quote. Or `swapsTokens` contains a token that has an `undefined` address.

```
const destinationTokenInfo =
      swapsTokens?.find(({ address }) => address === toTokenAddress) ||
      selectedToToken;
 ```
 
 It would be good if we could verify exactly how the request for quotes succeeds when `const selectedToToken = {}` and what the address is in that case.

darkwing replied:
We could remove the `{}` fallback from `selectedFromToken` and `selectedToToken` that existed from the initial swaps merge, and remove the `toTokenAddress` and `fromTokenAddress` checks in this PR.

If we should never get there without these properties, we should probably throw an error here, right?

danjm replied:
> If we should never get there without these properties, we should probably throw an error here, right?

Right, but what I want to know is how was it possible to request quotes without these properties? Requesting quotes when from or two are empty strings or undefined would return 0 quotes.

danjm replied:
I think this can be addressed in another PR, as the change to `actions.js` handles the objectives of this PR.

darkwing commented on ui/app/ducks/swaps/swaps.js:
```diff
@@ -442,6 +446,7 @@ export const fetchQuotesAndSetQuoteState = (
       );
     }
     if (
+      fromTokenAddress &&
```
Same here

darkwing commented on ui/app/store/actions.js:
```diff
@@ -1382,6 +1382,9 @@ export function addToken(
 ) {
   return (dispatch) => {
+    if (!address) {
+      throw new Error('MetaMask - Cannot add token without address');
```
Note to self: we should probably `reject` here.

darkwing commented:
@danjm Since you have the most knowledge about the swaps codebase, could you take a look at this?

danjm commented on app/scripts/migrations/053.js:
```diff
@@ -0,0 +1,23 @@
+ * Remove tokens that don't have an address due to
+ * lack of previous addToken validation
+ */
+export default {
```
Also, we should check if there could possibly undefined or null values in `accountHiddenTokens` and `hiddenTokens`

darkwing replied:
Added all of those new `...tokens` keys!

danjm commented:
To fully address #10593, should a change also be made on the 'Add Token' screen so that typing a single character does not result in an error state when there is a flawed token in the users data?

Gudahtt commented:
>when there is a flawed token in the users data?

Doesn't this PR prevent this circumstance? I don't quite understand what situation you're referring to.

danjm commented:
> Doesn't this PR prevent this circumstance? I don't quite understand what situation you're referring to
Right, but suppose such a token does somehow get added. The add token screen shouldn't error during typing right? I mean, add token functionality should continue to work regardless of what the data is like for already added tokens.

Gudahtt commented on ui/app/store/actions.js:
```diff
@@ -1382,6 +1382,11 @@ export function addToken(
   dontShowLoadingIndicator,
 ) {
   return (dispatch) => {
+    if (!address) {
```
This change alone should prevent adding tokens without an address.

If there's uncertainty about where in the UI this was made possible (i.e. whether it was in swaps, or the confirm suggested token, or both) we could ship just this change and the migration. Then we'd be preventing invalid state, and we'd get nice helpful Sentry errors for any remaining UI bugs.

darkwing commented on app/scripts/migrations/054.js:
```diff
@@ -0,0 +1,57 @@
+    // These properties are arrays and can be directly filtered
+    const tokenArrays = ['tokens', 'hiddenTokens'];
+    tokenArrays.forEach((tokenKey) => {
+      if (PreferencesController[tokenKey]) {
```
This `if` is required because neither `PreferencesController.tokens` or `PreferencesController. hiddenTokens` exists in the migration data to this point.

danjm replied:
What do you mean by "neither PreferencesController.tokens or PreferencesController. hiddenTokens exists in the migration data to this point"? If they don't exist, is this `forEach` necessary? Will the code inside the if block ever run?

Perhaps you meant that it is possible for PreferencesController.tokens or PreferencesController.hiddenTokens to not exist at this point.

darkwing commented on app/scripts/migrations/054.js:
```diff
@@ -0,0 +1,57 @@
+    // These properties are: (account) -> (network) -> [(token), ....]
+    const tokenObjects = ['accountHiddenTokens', 'accountTokens'];
+    tokenObjects.forEach((tokenObject) => {
+      if (PreferencesController[tokenObject]) {
```
This `if` is required because neither `PreferencesController. accountHiddenTokens ` or `PreferencesController. accountTokens ` exists in the migration data to this point.

darkwing commented on app/scripts/migrations/054.js:
```diff
@@ -0,0 +1,57 @@
+
+    // Removes an undefined key from "assetImages", which was added by addToken
+    if (
+      PreferencesController.assetImages &&
```
`assetImages` isn't in our migration data to this point.

darkwing commented:
@danjm One thing I notice is:

0.  Open the extension in full screen mode
1.  Click "Swap" from the home screen -- "ETH" is the selected "from" token
2. Refresh the page
3. No "From" token is selected.

We should probably default to ETH, since that's what the home screen does.  Maybe we do this in a separate PR?

brad-decker reviewed (commented):
LGTM, though would like @danjm to revisit his comment and subsequent changes before merge.

danjm commented on app/scripts/migrations/054.js:
```diff
@@ -0,0 +1,57 @@
+      if (PreferencesController[tokenKey]) {
+        PreferencesController[tokenKey] = PreferencesController[
+          tokenKey
+        ].filter(({ address }) => address);
```
`hiddenTokens` is an array of strings. If you inspect state, it looks like this:
```
"hiddenTokens": [
  "0x2260fac5e5542a773aa44fbcfedf7c193bc2c599"
],
```
So this filter would filter out all hidden tokens.

danjm commented on app/scripts/migrations/054.js:
```diff
@@ -0,0 +1,57 @@
+            PreferencesController[tokenObject][account][
+              network
+            ] = PreferencesController[tokenObject][account][network].filter(
+              ({ address }) => address,
```
The values pointed to by network ids in the `accountHiddenTokens` objects are arrays of strings. They look like this:
```
    "accountHiddenTokens": {
      "0xd2eab055e00d7daeffec4ef62b65bd069e204704": {
        "0x1": [
          "0x2260fac5e5542a773aa44fbcfedf7c193bc2c599"
        ],
        "0x539": [],
        "0x3": []
      }
    },
```
So this filter would always result in an empty array when `tokenObject === 'accountHiddenTokens'`

danjm reviewed (changes requested):
I left a couple comments on the migration that should be addressed

brad-decker reviewed (dismissed):
changes LGTM @danjm can you give this another pass, please 🙏🏻

brad-decker commented:
@darkwing you'll need to move your migration test into the migrations folder alongside 054.js, and then apply your changes to `test/unit/ui/app/ations.test.js` to `ui/app/store/actions.test.js`. Let me know if I can help.

darkwing committed:
Fix 10517 - Prevent tokens without addresses from being added to token list

darkwing commented:
Rebased per @brad-decker's directives.  I also removed a test that I had added, as since ....

```
const addTokenStub = sinon
        .stub()
        .callsFake((_, __, ___, ____, cb) => cb(new Error('error')));
```

... makes `addToken` fail regardless of arguments, it doesn't make sense to test that not providing a token address and pretend it represents anything.

darkwing commented:
@Gudahtt @danjm @brad-decker Since this is my first migration, I'd love a deep dive on it, especially test data approval.

brad-decker commented on app/scripts/migrations/054.js:
```diff
@@ -0,0 +1,71 @@
+    if (PreferencesController.accountHiddenTokens) {
+      Object.keys(PreferencesController.accountHiddenTokens).forEach(
+        (address) => {
+          const networks = Object.keys(
```
upon deeper dive, it might be slightly more correct to name this `chains` instead of `networks` because the keys are things like `0x1` versus `mainnet`. This is *purely* for the readability of the migration.

brad-decker reviewed (changes requested):
I walked through and ran this test a few times, I only have one minor NIT. I think this is a fine migration and test/data set.

darkwing commented:
Good idea Brad, updated!  I also added `Array.isArray` checks for increased type checking security.

brad-decker reviewed (dismissed):
LGTM

darkwing committed:
Avoid touching hiddenTokens and accountHiddenTokens as they were not affected

darkwing commented:
In the interest of keeping migrations as tight as possible, I've removed checks on `hiddenTokens` and `accountHiddenTokens`, as a review of the affected user's state logs show that those properties weren't affected.

Gudahtt reviewed (approved):
LGTM!

danjm reviewed (approved):
LGTM
